src/async_wrap.cc

```diff
@@ -254,16 +259,24 @@ static void PromiseHook(PromiseHookType type, Local<Promise> promise,
       wrap = PromiseWrap::New(env, promise, parent_wrap, silent);
     } else {
       wrap = PromiseWrap::New(env, promise, nullptr, silent);
+
+      // needed for async functions :/
+      // the top level will not emit before and after
+      env->async_hooks()->push_execution_async_resource(wrap->object());
```

Comment:
I don't understand why an initialization of a Promise should add a resource on stack which gets removed once it's resolved. Maybe it's just missing knowledge on how promise hooks work.
Maybe a test using plain promises instead `await` should be added?

Reply:
An async function has an outer promise for the function itself and inner promises between each await. The outer promise doesn’t actually trigger before/after promise hook events so in that special case we wrap with init and resolve instead. If we didn’t there’d be a blank spot between the start of the function and the first await where there is no resource.

Reply:
But what about a pure `new Promise()` call without any async function involved? Wouldn't this end up in an unwanted push?

Reply:
There's no before/after with `new Promise(...)` either. Only `then(...)` or `await` produces before/after.

Reply:
Following test fails but I think it should not:
```
const ah = require("async_hooks")
const assert = require("assert");

ah.createHook({ init() {}}).enable();
const er = ah.executionAsyncResource();
new Promise((res, rej) => {});
assert.strictEqual(ah.executionAsyncResource(), er);
```

Reply:
Hmm...yeah, because of popping the initial promise resource on resolve, the rest of the sync stack after the init will have the promise at the top of the stack rather than the original resource from that tick. It'd be nice if v8 gave us before/after events around the executor function for `new Promise` or the initial part of async functions. :thinking: 

Not too sure what to do here. Could maybe just _not_ use the promise as the resource for the executor and document that possibly unexpected case. :man_shrugging:

Reply:
I still haven't understood why push on init and pop on resolve is needed. My understanding is that the current executing resource in the sync promise executor is the caller of `new Promise()`. 
Do you have some sample which shows the wrong behavior you have in mind if you remove the two lines?

Reply:
Yes, the resource in the `new Promise(...)` executor is the caller of `new Promise(...)` rather than the promise itself. This is maybe less controversial in the constructor form, but in the async function form this means that the bit in an async function before the first await has a resource of the caller of the async function rather than the async function itself. This means that if you build a local storage concept on top of async_hooks and try to store something to be accessible within the scope of the async function it'll actually leak out to the upper context after executing the function.

Reply:
Honestly speaking this is exactly what I would expect. if you set something to your AsyncLocal storage it follows the flow across async boundaries - but it is also available on your current sync path.
Why should this differ for async functions compare to sync?
If the called function represents some new transaction path which should get a new context then the init hook (or the instrumenation in the function) has to detect this and cause a fork of the AsyncLocal storage by creating a new context and propagating this one instead that one active. But this is mostly the same for sync functions with a callback as argument.

Reply:
Because it's not clear that the first part of an async function is sync. Most people assume the whole thing is just encapsulated in a promise. See this example:

```js
'use strict'

const { createHook, executionAsyncResource } = require('async_hooks')

const contexts = new WeakMap()

function cls () {
  const resource = executionAsyncResource()
  if (!contexts.has(resource)) contexts.set(resource, {})
  return contexts.get(resource)
}

createHook({
  init(asyncId, type, triggerAsyncId, resource) {
    const context = {}
    Object.setPrototypeOf(context, cls())
    contexts.set(resource, context)
  }
}).enable()

function sleep () {
  return new Promise(setImmediate)
}

cls().i = 0

setImmediate(() => {
  console.log(cls().i) // 1, because main() leaks until the first await
})

async function main() {
  cls().i++
  await sleep()
  cls().i++
  await sleep()
  cls().i++
}

main()
```

Reply:
So … one way or another, I feel like the resource should *always* match the current set of async ids, right? i.e. anywhere we call `push_async_ids()` and `pop_async_id()` or their JS equivalents, we should also push and pop the resource object?

Reply:
I decided to remove the resource push/pop in the PromiseHook handler, as you suggested. It seems like it's perhaps the _less broken_ option, but I _do_ still feel the most _correct_ thing would be for the promise executor and the first part of an async function to trigger before/after events to allow the resource specifically for that short period of time, even if it _is_ sync, to be attributed to the promise rather than the caller.

Reply:
Yeah, but whatever you do here, please make sure to also make sure that the reported async ids match that…

Reply:
Yep, so after removing the PromiseHook bits they should now always match exactly, other than the top-level having an empty object resource, since there's no other handle or request to use there. Also, if we can get a _proper_ before/after around the promise executor we'll get the correct ids and resources automatically. :)

Reply:
I agree that a before/after for the promise executor/first part of async function would be nice. But I would like to see them same also for non promise resources.
For example some `db.query(query, cb) => void` creating a new async operation should look equal to the corresponding promisifed variant `db.query(query) => Promise`.
Exceptions are low level operations like `nextTick`,... where the sync part is not really of interest to track/monitor - and they don't create further child async operations.

For now I like the solution as it is.

Reply:
I think this is resolved then? If you disagree, feel free to switch this back to unresolved.